src/support/safe_math.h

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_safe_math_h
+#define wasm_safe_math_h
+
+namespace wasm {
+
+#if !__has_builtin(__builtin_add_overflow)
+template<typename T> bool __builtin_add_overflow(T a, T b, T* output) {
```

Comment:
It would be better to give this a different name and implement it in terms of `__builtin_add_overflow` if that exists like LLVM does. That avoids both UB (using a double underscore prefix) and prevents the failure mode where we could forget to include this header and only discover the problem on CI (if at all) because the local compiler has this builtin.

Reply:
Hmm, I'm not sure I would call using raw `__builtin*` functions idiomatic, though. If we really want to be idiomatic here, we could use `<stdckdint.h>` from C23. We could import LLVM's implementation, for instance.

Reply:
Interesting, thanks, I hadn't heard about `stdckdint` before. Looking at that, it seems it isn't fully standardized yet (even though 2023 is in the past? but maybe online links are obsolete), but more importantly I do not see such a file in `llvm-project` anywhere, so I suspect it isn't implemented there yet. I did a little searching but can't find a permissively-licensed one elsewhere.

I think it's reasonable to consider `__builtin_add_overflow` as idiomatic given all modern compilers implement it. It seems to be the best "standard" here atm.

Reply:
Here it is: https://github.com/llvm/llvm-project/blob/3694697003bbf443fd644c6746a1c7c937657fce/clang/lib/Headers/stdckdint.h

I think it's simple enough that it's worth using.

Reply:
Thanks - looks like I missed it because that file is 3 months old and my LLVM checkout is older than that :smile:  I only update when I need to bisect.

Looks like that is really just 3 lines:
```c
#define ckd_add(R, A, B) __builtin_add_overflow((A), (B), (R))
#define ckd_sub(R, A, B) __builtin_sub_overflow((A), (B), (R))
#define ckd_mul(R, A, B) __builtin_mul_overflow((A), (B), (R))
```
Are you saying to copy those lines, use those macros in user code, and use the existing code in this PR as a polyfill? If so, what header would user code include (my existing one with new contents, or a bundled/vendored but modified new file by name `stdckdint.h`)? Sorry, I'm a little unsure what you mean here.

Reply:
I think we should vendor the `stdckdint.h`, at least until compilers support it on their own. That seems like the most future-proof solution IMO.

Reply:
To be sure I understand you, after that vendoring we still need the polyfill I started here (since that LLVM file adds no support for old compilers). Where are you suggesting that polyfill be - in that vendored file, or a file that wraps it somehow, or something else?

Reply:
Sounds good. In that case there's really nothing useful in the upstream file, so I just converted my polyfill to have that name and use the C23 name for the API, PTAL.